### PR TITLE
Spectator changes + fix performance bug

### PR DESCRIPTION
**Pull request by flux627**
Forked from @altf4's `spectator` branch [here](https://github.com/altf4/slippi-desktop-app/tree/spectator) which should be used with [this corresponding Ishiiruka PR](https://github.com/project-slippi/Ishiiruka/pull/98).

Includes all of his changes, along with:

- Refactor `SlpFileWriter#currentFile` to instance of new `SlpFile` class. (Open to better names for this new class.)
- Only merge `previousBuffer` if it exists
- Properly check if clients array has items with `.length`
- Only concat buffers when `fullBuffer` is accessed via getter
- Linting changes / fixes

This PR defers merging of buffer data to `fullBuffer` until it is accessed, which resolves performance issues *as long as no clients are connected via relay mode*. This problem still exists in relay mode, but will be addressed in a subsequent PR.

### CODE REVIEW

**Comment by altf4** [comment]
Commit a774606 breaks the functionality. I disabled the `connectOnPort()` functions because they didn't play nicely with the changes I was making.

I think what happens is that the ENet port connects just fine, but then the program tries to also connect on the two TCP ports. These fail, and it changes the connection status, breaking the existing connection. I'm not sure what the best way to handle this is? The behavior we want I guess is to try all three connection ports and accept the first one that works?

File: app/domain/ConsoleConnection.js [hunk 1]
```diff
@@ -135,18 +138,107 @@ export default class ConsoleConnection {
...
     this.dolphinManager.updateSettings(connectionSettings);
 
-    if (this.port && this.port !== Ports.WII_LEGACY && this.port !== Ports.WII_DEFAULT) {
-      // If port is manually set, use that port. Don't do this if the port is set to legacy as
-      // somebody might have accidentally set it to that and they would encounter issues with
-      // the new Nintendont
-      this.connectOnPort(this.port);
-      return;
-    }
+    // if (this.port && this.port !== Ports.WII_LEGACY && this.port !== Ports.WII_DEFAULT) {
```
Comment (a-sink-a-wait) [inline app/domain/ConsoleConnection.js:141]: Commented code?

Comment (altf4) [reply app/domain/ConsoleConnection.js:141]: I meant these two lines:

https://github.com/project-slippi/slippi-desktop-app/blob/d7db6b69bcc875a20ba93ba0e019c5975995c49b/app/domain/ConsoleConnection.js#L150

**Comment by flux627** [comment]
This PR still has problems, but I probably won't have time to work on it more until this weekend. If anyone else wants to carry the torch forward before then, please do so by all means.

**Comment by altf4** [comment]
When trying to resume in the middle of a match (>4 mins long) I get this error:

image

Which apparently is caused by this?

https://stackoverflow.com/questions/27769842/write-after-end-error-in-node-js-webserver

**Comment by NikhilNarayana** [comment]
closing in favor of the https://github.com/project-slippi/slippi-desktop-app/tree/feature/broadcast branch

**Pull request closed without merging**